### ISSUE TITLE
Question about embedding

### ISSUE THREAD

**Issue by kpmokpmo** [QData/spacetimeformer#1]
Hi, great job!

I just have a quick question about the embedding part. Although position embedding is a tradition in Transformer based models, it seems you embeded so much different kinds of information into the original sequence in an addition way. I am not sure if my understanding is correct. The concatenation form is easier for the network to learn at cost of more parameters and vice versa for the addition form. I believe the nerual network bears certain capability to extract different information from the raw input and projects them into higher order latent space, but would that possibly deteriorates the performance if we add too much? And will large hidden dim mitigates?

BTW, I also heard of the explanation: (a+b)' = a' + b' which means the addition and concatenation operations are equal fom a gradient perspective. And that contradicts to my gut feeling.

Thanks very much!

**Comment by jakegrigsby** [comment]
Yes this was definitely something I considered. We actually started off by concatenating separate (smaller) Time2Vec, Value and Variable embeddings. That worked okay from what I remember, but so many other parts of the model have changed since then that the performance is hard to compare. The main reason we switched to the summation method is because it gets rid of the hyperparameters that decide how long each concatenated vector should be. It was difficult to say whether the Time vector should have more elements than the Variable vector, for example. 

I think this is related to your explanation, but the way I look at it is the summation of learnable embeddings is a generalization of concatenation. If concatenation is really the best option, the layers that generate the embeddings could learn a sparse output that looks a lot like concatenation after summation. I'm honestly not sure whether anything like that happens in practice... but it's possible, and so are all the embeddings that concatenation could not represent.

Most Transformer models learn the sum of two embeddings (value and position). The Val+Time embedding is very similar to regular transformer projections except that we add additional Time2Vec time features. The variable embedding is similar to standard position embeddings except it is repeated for several tokens. We do add an extra "Given" embedding but it is not super important in practice. So our model doesn't have too much more information to interpret than the typical Transformer.

**Comment by kpmokpmo** [comment]
Well, I think there is a little theoretical information loss in the summation method. Say the raw sequence is 1,2,3 and a learned time embedding is 10,11,12. Then the final sequence received is 11,13,15。Although the nn knows the embedding code is an order of magnitude higher than the raw input, so 15 probably means "late in time", and it will minus the known embedding code to figure out the original data(thank god embeddings are learned but fixed which means they are known for sure), they still cannot do the seperation job accurately especially when we need to embed information from different modalities. And a high dimensional space seems to be a necessity for such a process.

**Comment by jakegrigsby** [comment]
That's true, but in that case the concatenation of embeddings is a higher dimension (2d) than the transformer input (1d) (`d_model` in the code). Some information has to be lost. But the size of concatenated embeddings is always <= `d_model` in practice, or we couldn't fit the tokens in the Transformer. If `d_model` is going to be 4, then we need to pick sizes for the value and variable embeddings so that they concatenate to 4 elements. If the value emb for one token is [1, 2] and the variable emb is [3, 4], the concatenated embedding of [1, 2, 3, 4] is the summation of embeddings [1, 2, 0, 0] and [0, 0, 3, 4]. So the embeddings we learn can recover concatenations of smaller vectors with total dimension <= `d_model`. We also don't have to specify the sizes of the value and variable embeddings in advance. If we think we need more dimensions for the embedding than we can fit in the model we need to find a way to lose some information. The concatenated method would have to chop off elements of some of its sub-embeddings to fit in `d_model`.  The method we use can try to learn a better way to condense the dimensions and the transformer been shown to be good at sorting out the information later.

Maybe a shorter explanation is that any concatenation of embeddings with total size `d_model` can be learned by the summation of `d_model` embeddings, and we can learn more complex combinations of the features when space is limited. I hope that makes sense. Thanks for checking out our work!

**Comment by kpmokpmo** [comment]
Hi, sorry for late response. I'm agree with you. It's essentially a trade-off between accuracy and throughput. I tried to think about it in an intuitive way: for a classification or univariate prediction problem with multivariate input or covariates, we would never like to squeeze the n-dim series into one by summation. Despite the fact that learnable absolute embedding is fixed in a forward process, the adaptive relative embedding plays a quite simlilar role as the time series information from one more sensor.  Deep learning is an experimental science anyway, so we got to find a balancing point. 

Thank you again for your great work and detailed explanation！
